File: workers/base/hive_base/joblog.py
"""Per-job logger: publishes live to Redis pub/sub + buffers for batched Postgres inserts."""
from __future__ import annotations
import asyncio
import json
from datetime import datetime, timezone
from typing import Any, Optional
import redis.asyncio as redis_async
from . import db as dbmod

FLUSH_BATCH = 50


class JobLogger:
    """One JobLogger per job. Use .info()/.warn()/.error()/.debug() then .flush() at end."""

    def __init__(
        self,
        *,
        job_id: str,
        redis_client: redis_async.Redis,
        dsn: str,
        batch_size: int = FLUSH_BATCH,
    ) -> None:
        self.job_id = job_id
        self._redis = redis_client
        self._dsn = dsn
        self._batch_size = batch_size
        self._buffer: list[dict[str, Any]] = []
        self._lock = asyncio.Lock()
        self._channel = f"hive:logs:{job_id}"

    async def _emit(self, level: str, message: str, meta: Optional[dict[str, Any]] = None) -> None:
        ts = datetime.now(timezone.utc)
        payload = {
            "ts": ts.isoformat(),
            "level": level,
            "message": message,
            "meta": meta,
        }
        # Live publish.
        try:
            await self._redis.publish(self._channel, json.dumps(payload, default=str))
        except Exception:
            # Logging must never crash the job; swallow.
            pass
        # Buffer for Postgres.
        row = {
            "id": dbmod.cuid_like(),
            "jobId": self.job_id,
            "level": level,
            "message": message,
            "meta": meta,
            "timestamp": ts,
        }
        async with self._lock:
            self._buffer.append(row)
            if len(self._buffer) >= self._batch_size:
                await self._flush_locked()

    async def _flush_locked(self) -> None:
        if not self._buffer:
            return
        rows = self._buffer
        self._buffer = []
        try:
            await dbmod.insert_logs(self._dsn, rows)
        except Exception:
            # Don't lose them — put them back at the head; but cap to avoid runaway.
            self._buffer = (rows + self._buffer)[:1000]

    async def flush(self) -> None:
        async with self._lock:
            await self._flush_locked()

    async def info(self, message: str, **meta: Any) -> None:
        await self._emit("info", message, meta or None)

    async def warn(self, message: str, **meta: Any) -> None:
        await self._emit("warn", message, meta or None)

    async def error(self, message: str, **meta: Any) -> None:
        await self._emit("error", message, meta or None)

    async def debug(self, message: str, **meta: Any) -> None:
        await self._emit("debug", message, meta or None)

    async def signal_terminal(self, status: str) -> None:
        """Tell SSE subscribers the job is done so they can close cleanly."""
        try:
            await self._redis.publish(
                self._channel,
                json.dumps({"__terminal": True, "status": status}),
            )
        except Exception:
            pass

File: packages/worker-base-ts/src/joblog.ts
import { Redis } from 'ioredis';
import { prisma, Prisma } from '@hive/db';
import { STREAMS } from './lifecycle.js';

export type LogLevel = 'debug' | 'info' | 'warn' | 'error';

export interface LogEntry {
  ts: string;
  level: LogLevel;
  message: string;
  meta?: Record<string, unknown> | null;
}

const FLUSH_BATCH = 50;

interface BufferedRow {
  jobId: string;
  level: LogLevel;
  message: string;
  meta: Prisma.InputJsonValue | typeof Prisma.JsonNull;
  timestamp: Date;
}

export class JobLogger {
  private readonly jobId: string;
  private readonly redis: Redis;
  private readonly channel: string;
  private readonly batchSize: number;
  private buffer: BufferedRow[] = [];
  private flushing = false;

  constructor(jobId: string, redis: Redis, batchSize: number = FLUSH_BATCH) {
    this.jobId = jobId;
    this.redis = redis;
    this.channel = STREAMS.logs(jobId);
    this.batchSize = batchSize;
  }

  async info(message: string, meta?: Record<string, unknown>): Promise<void> {
    await this.emit('info', message, meta);
  }
  async warn(message: string, meta?: Record<string, unknown>): Promise<void> {
    await this.emit('warn', message, meta);
  }
  async error(message: string, meta?: Record<string, unknown>): Promise<void> {
    await this.emit('error', message, meta);
  }
  async debug(message: string, meta?: Record<string, unknown>): Promise<void> {
    await this.emit('debug', message, meta);
  }

  private async emit(level: LogLevel, message: string, meta?: Record<string, unknown>): Promise<void> {
    const ts = new Date();
    const payload: LogEntry = {
      ts: ts.toISOString(),
      level,
      message,
      meta: meta ?? null,
    };
    try {
      await this.redis.publish(this.channel, JSON.stringify(payload));
    } catch {
      /* live logs are best-effort */
    }
    this.buffer.push({
      jobId: this.jobId,
      level,
      message,
      meta: (meta ?? Prisma.JsonNull) as Prisma.InputJsonValue | typeof Prisma.JsonNull,
      timestamp: ts,
    });
    if (this.buffer.length >= this.batchSize) {
      void this.flush();
    }
  }

  async flush(): Promise<void> {
    if (this.flushing || this.buffer.length === 0) return;
    this.flushing = true;
    const rows = this.buffer;
    this.buffer = [];
    try {
      await prisma.jobLog.createMany({ data: rows });
    } catch (err) {
      this.buffer = rows.concat(this.buffer).slice(0, 1000);
      throw err;
    } finally {
      this.flushing = false;
    }
  }

  async signalTerminal(status: 'succeeded' | 'failed' | 'cancelled'): Promise<void> {
    try {
      await this.redis.publish(this.channel, JSON.stringify({ __terminal: true, status }));
    } catch {
      /* ignore */
    }
  }
}
